Userland/Libraries/LibGfx/ImageFormats/WebPLoaderLossy.cpp

```diff
@@ -490,15 +494,430 @@ ErrorOr<Vector<MacroblockMetadata>> decode_VP8_macroblock_metadata(BooleanDecode
+    //  tables (the tables are given in Section 14)."
+    // Section 14 then lists two (!) fixed tables (which are in WebPLoaderLossyTables.h)
+
+    // "Lookup values from the above two tables are directly used in the DC
+    //  and AC coefficients in Y1, respectively.  For Y2 and chroma, values
+    //  from the above tables undergo either scaling or clamping before the
+    //  multiplies.  Details regarding these scaling and clamping processes
+    //  can be found in related lookup functions in dixie.c (Section 20.4)."
+    // Apparently spec writing became too much work at this point. In section 20.4, in dequant_init():
+    // * For y2, the output (!) of dc_qlookup is multiplied by 2, the output of ac_qlookup is multiplied by 155 / 100
+    // * Also for y2, ac_qlookup is at least 8 for lower table entries (XXX!)
+    // * For uv, the dc_qlookup index is clamped to 117 (instead of 127 for everything else)
+    //   (or, alternatively, the value is clamped to 132 at most)
+
+    u8 y_ac_base = quantization_indices.y_ac;
+    if (segmentation.update_macroblock_segmentation_map) {
+        if (segmentation.segment_feature_mode == SegmentFeatureMode::DeltaValueMode)
+            y_ac_base += segmentation.quantizer_update_value[segment_id];
+        else
+            y_ac_base = segmentation.quantizer_update_value[segment_id];
+    }
+
+    u8 dequantization_index;
+    if (index.is_y2())
+        dequantization_index = y_ac_base + (is_dc ? quantization_indices.y2_dc_delta : quantization_indices.y2_ac_delta);
+    else if (index.is_u() || index.is_v())
+        dequantization_index = y_ac_base + (is_dc ? quantization_indices.uv_dc_delta : quantization_indices.uv_ac_delta);
+    else
+        dequantization_index = is_dc ? (y_ac_base + quantization_indices.y_dc_delta) : y_ac_base;
+
+    // clamp index
+    if ((index.is_u() || index.is_v()) && is_dc)
+        dequantization_index = min(dequantization_index, 117);
+    else
+        dequantization_index = min(dequantization_index, 127);
+
+    // "the multiplies are computed and stored using 16-bit signed integers."
+    i16 dequantization_factor;
+    if (is_dc)
+        dequantization_factor = (i16)dc_qlookup[dequantization_index];
+    else
+        dequantization_factor = (i16)ac_qlookup[dequantization_index];
+
+    if (index.is_y2()) {
+        if (is_dc)
+            dequantization_factor *= 2;
+        else
+            dequantization_factor = (dequantization_factor * 155) / 100;
+    }
+
+    return dequantization_factor * value;
+}
+
+// Reading macroblock coefficients requires needing to know if the block to the left and above the current macroblock
+// has non-zero coefficients. This stores that state.
+struct CoefficientReadingContext {
+    // Store if each plane has nonzero coefficients in the block above and to the left of the current block.
+    Vector<bool> y2_above;
+    Vector<bool> y_above;
+    Vector<bool> u_above;
+    Vector<bool> v_above;
```

Comment:
Can we make these `AK::Bitmaps`?

Reply:
There are `ceil_div(width * 9, 16)` of these, so 576 bytes for a 1024 wide image.

A 1024-wide image needs 4096 bytes to store the first scanline. Assuming it's just 10 pixel high, the 576 * 7/8 = 504 bytes savings are ~1% of the size needed for the image, and this is transient memory, only used during decoding.

So not sure this is worth worrying about.